kernel/kernel_drivers.go

```diff
@@ -0,0 +1,208 @@
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package kernel
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+)
+
+// For testing purposes
+var osSymlink = os.Symlink
+
+// We expect as a minimum something that starts with three numbers
+// separated by dots for the kernel version.
+var utsRelease = regexp.MustCompile(`^([0-9]+\.){2}[0-9]+`)
+
+// KernelVersionFromModulesDir returns the kernel version for a mounted kernel
+// snap (this would be the output if "uname -r" for a running kernel). It
+// assumes that there is a folder named modules/$(uname -r) inside the snap.
+func KernelVersionFromModulesDir(mountPoint string) (string, error) {
+	modsDir := filepath.Join(mountPoint, "modules")
+	entries, err := os.ReadDir(modsDir)
+	if err != nil {
+		return "", err
+	}
+
+	kversion := ""
+	for _, node := range entries {
+		if !node.Type().IsDir() {
+			continue
+		}
+		if !utsRelease.MatchString(node.Name()) {
+			continue
+		}
+		if kversion != "" {
+			return "", fmt.Errorf("more than one modules directory in %q", modsDir)
+		}
+		kversion = node.Name()
+	}
+	if kversion == "" {
+		return "", fmt.Errorf("no modules directory found in %q", modsDir)
+	}
+
+	return kversion, nil
+}
+
+func createFirmwareSymlinks(orig, dest string) error {
+	fwOrig := filepath.Join(orig, "firmware")
+	fwDest := filepath.Join(dest, "lib", "firmware")
+	if err := os.MkdirAll(fwDest, 0755); err != nil {
+		return err
+	}
+
+	// Symbolic links inside firmware folder - it cannot be directly a
+	// symlink to "firmware" as we will use firmware/updates/ subfolder for
+	// components.
+	entries, err := os.ReadDir(fwOrig)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// Bit of a corner case, but maybe possible. Log anyway.
+			logger.Noticef("no firmware found in %q", fwOrig)
+			return nil
+		}
+		return err
+	}
+
+	for _, node := range entries {
+		lpath := filepath.Join(fwDest, node.Name())
+		origPath := filepath.Join(fwOrig, node.Name())
+		switch node.Type() {
+		case 0, fs.ModeDir:
+			// Create link for regular files or directories
+			if err := os.Symlink(origPath, lpath); err != nil {
+				return err
+			}
+		case fs.ModeSymlink:
+			// Replicate link (it should be relative)
```

Comment:
should we check that is relartive?  should we also check it when we pack a kernel snap as well?

Reply:
I've added a TODO for snap pack.

Reply:
Added with a test now